File: Homework/Homework Assignment 2/Codes/cakar_berk.dart
void main() {
  print("INTRODUCTION TO USER-LOCATED LOOP CONTROL MECHANISMS IN DART");

  while (true) {
    print("iterating once");
    break;
  }

  // ignore: dead_code
  for (int i = 0; i < 10; i++) {
    continue; // continue to the next iteration without executing the rest of the loop body (print statement)
    // ignore: dead_code
    print("$i");
  }

  // Labels can also be used to break out of nested loops (see the examples below)

  print("\nSAMPLE 1: BREAK -------------------------------------------------------------");

  // An example for break
  for (int i = 0; i < 44; i++) {
    if (i == 2) {
      /*
         When i is 2 in the outer loop, break statement will be executed
         and the outer loop will be terminated
      */
      print(
          "  in outer loop i is 2, now break out of the outer loop, otherwise i would go as 0, 1, 2, ..., 43");
      break;
    }

    print("in outer loop i is $i");

    for (int j = 0; j < 12; j++) {
      /*
         When j is 5 in the inner loop, break statement will be executed
         However, the outer loop will continue to run since the break statement
         only breaks out of the inner loop
      */
      if (j == 3) {
        print(
            "  in inner loop j is 3, now break out of the inner loop, otherwise j would go as 0, 1, 2, ..., 11");
        break;
      }
      print("  in inner loop j is $j");
    }
  }

  print("\nSAMPLE 2: CONTINUE ----------------------------------------------------------");

  // An example for continue
  var i = 0;
  while (i < 10) {
    i++;
    if (i % 2 == 0) {
      /*
         When i is even, continue statement will be executed
         and the while loop will continue to the next iteration
         without executing the print statement for even numbers
      */
      continue;
    }
    print("in while loop, i is odd, now printing: i = $i");
  }

  print("\nSAMPLE 3: LABELED BREAK -----------------------------------------------------");

  // An example for labeled break
  outerLoop:
  for (int j = 0; j < 10; j++) {
    print("in outer loop j is $j");

    int i = 0;
    // In Dart, labels can be used to break or continue a specific loop
    innerLoop:
    do {
      print("  in inner loop, i is $i");
      /*
         When j is 5, break statement will be executed in the inner loop
         and the outer loop will be terminated. Otherwise, the outer loop
         would continue to run until j is 9
      */
      if (j == 5) {
        print("j is 5, break out the outer loop, normally 'j' would go as 0, 1, 2, ..., 9");
        break outerLoop;
      }
      /*
         When i is 3, break statement will terminate the inner loop and
         the outer loop will continue to run. Otherwise, the inner loop
         would continue to run until i is 2
      */
      if (i == 1) {
        print("  i is 1, break out the inner loop, normally 'i' would go as 0, 1, 2");
        break innerLoop;
      }
      i++;
    } while (i < 3);
  }

  print("\nSAMPLE 4: LABELED CONTINUE --------------------------------------------------");

  // An example for labeled continue
  outerLoop:
  for (var i = 0; i <= 3; i++) {
    print("in outer loop i is $i");
    innerLoop:
    for (var j = 0; j <= 2; j++) {
      /*
         When j is 1, continue statement will be executed in the inner loop
         which means the inner loop will continue to the next iteration
         without executing the print statement when j is 1
      */
      if (j == 1) {
        continue innerLoop;
      }
      /*
         When i is 2, continue statement will be executed in the outer loop
         which means the outer loop will continue to the next iteration
         with skipping the inner loop
      */
      if (i == 2) {
        continue outerLoop; // same as break innerloop;
      }
      print("  in inner loop j is $j");
    }
  }
}

File: Homework/Homework Assignment 2/Codes/cakar_berk.lua
print("INTRODUCTION TO USER-LOCATED LOOP CONTROL MECHANISMS IN LUA")

while true do
    print("iterating once")
    break
end

for i = 1, 10 do
    -- continue is not directly supported in Lua
    goto continue -- continue to the next iteration without executing the rest of the loop body (print statement)
    print("iterating once")
    ::continue::
end

print("\nSAMPLE 1: BREAK -------------------------------------------------------------");

-- An example for break
-- In this for loop, a modified Floyd's triangle is printed
for i = 1, 10 do
    for j = 1, i do
        io.write("*")
        if j == 5 then
            -- When the number of stars printed (columns) is 5, break out of the inner loop
            io.write(" The inner loop has completed its fifth iteration (column), now break out the inner loop (stop adding new columns)")
            break
        end
    end
    if i == 8 then
        -- When the number of rows is 8, break out of the outer loop
        io.write("\nThe outer loop has completed its eight iteration (row), now break out the outer loop (stop adding new rows)")
        break
    end
    print();
end
-- In the end, it is expected to see a trapazoid with 8 rows and 5 columns

-- Lua does not have a continue statement, but you can use the goto statement to jump to the next iteration of a loop.

io.write("\n\nSAMPLE 2: LABELED BREAK (VIA GOTO) ---------------------------------------\n");

-- An example for labeled break
local array = {
    { 1,  2,  3,  4,  5},
    { 6,  7,  8,  9, 10},
    {11, 12, 13, 14, 15},
    {16, 17, 18, 19, 20},
    {21, 22, 23, 24, 25}
}
for i = 1, #array do
    for j = 1, #array[i] do
        if i == 4 then
            -- When the fourth row of the 2D array is reached, break out of the outer loop
            goto outerLoopDone -- break out of the outer loop
        end
        if j == 4 then
            -- When the fourth column of the 2D array is reached, break out of the inner loop
            goto innerLoopDone -- break out of the inner loop
        end
        io.write(array[i][j], "\t")
    end
    ::innerLoopDone::
    io.write("\n")
end
::outerLoopDone::
-- In the end, it is expected to see the left-top 3x3 sub-matrix of the 5x5 2D array

io.write("\nSAMPLE 3: LABELED CONTINUE (VIA GOTO) ---------------------------------------");

-- An example for labeled continue
for i = 1, 4 do
    io.write("\n")
    for j = 1, 3 do
        if i == j then
            -- When the row number equals the column number (i.e., a square number is encountered),
            -- continue the outer loop, so that no number is printed after the square number
            io.write(">", "\t")
            goto continue -- same as continue outerLoop, or break innerLoop
        end
        io.write(i * j, "\t")
    end
    ::continue::
end
print("\n> means i == j, which indicates a square number has been reached, so we continue to the next iteration of the outer loop")

